xdsl/dialects/func.py

```diff
@@ -47,6 +51,64 @@ def verify_(self) -> None:
                 "input types"
             )
 
+    @classmethod
+    def parse(cls, parser: Parser) -> FuncOp:
+        # Parse visibility keyword if present
+        visibility = parser.tokenizer.next_token_of_pattern(
+            re.compile("public|nested|private")
+        )
+        if isinstance(visibility, Span):
+            visibility = visibility.text
+        else:
+            visibility = "private"
+
+        # Parse function name
+        name = parser.parse_symbol_name().data
+
+        # Parse function arguments
+        args = parser.parse_comma_separated_list(
+            parser.Delimiter.PAREN,
+            lambda: parser.parse_optional_argument() or parser.parse_type(),
+        )
+
+        # Check consistency (They should be either all named or none)
+        if isa(args, list[parser.Argument]):
+            entry_args = args
+            input_types = [a.type for a in args]
+        elif isa(args, list[Attribute]):
+            entry_args = None
+            input_types = args
+        else:
+            parser.raise_error(
+                "Expected all arguments to be named or all arguments to be unnamed."
+            )
+
+        # This is for the args is list[parser.Argument] case
+        # Then input_types = [a.type for a in args]
+        # and Argument.type is Attribute | None
+        # But here we use parse_optional_argument(True), so those should never be None
+        # -> Should we have a different method for typing reasons?
+        if not isa(input_types, list[Attribute]):
+            parser.raise_error(
+                "Expected all arguments to be named or all arguments to be unnamed."
+            )
```

Comment:
@math-fehr

Reply:
Wrote a comment above, that should remove the need for this.

Reply:
Except parse_optional_type is not a thing :( I can go with try_parse_type, but am not sure about the lexer/tokenizer status there.
I'll just use try_parse_type meanwhile unless it's an issue!